File: plugins/background-location/android/src/main/java/com/johnnyoutlaw/unwaveringband/backgroundlocation/BackgroundLocationPlugin.java
package com.johnnyoutlaw.unwaveringband.backgroundlocation;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
    name = "BackgroundLocation",
    permissions = {
        @Permission(
            strings = {
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION
            },
            alias = "location"
        ),
        @Permission(
            strings = { Manifest.permission.ACCESS_BACKGROUND_LOCATION },
            alias = "background"
        ),
        @Permission(
            strings = { Manifest.permission.POST_NOTIFICATIONS },
            alias = "notifications"
        )
    }
)
public class BackgroundLocationPlugin extends Plugin {
    private SessionStore store;

    @Override
    public void load() {
        store = new SessionStore(getContext());
        if (store.isTrackingEnabled() && store.hasSession()) {
            startService();
        }
    }

    @PluginMethod
    public void setAuthSession(PluginCall call) {
        String access = call.getString("accessToken");
        String userId = call.getString("userId");
        String url = call.getString("supabaseUrl");
        String anon = call.getString("supabaseAnonKey");
        if (access == null || userId == null || url == null || anon == null) {
            call.reject("accessToken, userId, supabaseUrl, and supabaseAnonKey are required");
            return;
        }
        store.saveSession(access, call.getString("refreshToken"), userId, url, anon);
        JSObject ret = new JSObject();
        ret.put("stored", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void clearAuthSession(PluginCall call) {
        stopService();
        store.clearSession();
        JSObject ret = new JSObject();
        ret.put("cleared", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        if (Build.VERSION.SDK_INT >= 33
            && ContextCompat.checkSelfPermission(getContext(), Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
            requestPermissionForAlias("notifications", call, "permCallback");
            return;
        }
        if (!hasFineLocation()) {
            requestPermissionForAlias("location", call, "permCallback");
            return;
        }
        if (Build.VERSION.SDK_INT >= 29 && !hasBackgroundLocation()) {
            requestPermissionForAlias("background", call, "permCallback");
            return;
        }
        resolvePermissionStatus(call);
    }

    @PermissionCallback
    private void permCallback(PluginCall call) {
        if (!hasFineLocation()) {
            resolvePermissionStatus(call);
            return;
        }
        if (Build.VERSION.SDK_INT >= 29 && !hasBackgroundLocation()) {
            requestPermissionForAlias("background", call, "backgroundCallback");
            return;
        }
        resolvePermissionStatus(call);
    }

    @PermissionCallback
    private void backgroundCallback(PluginCall call) {
        resolvePermissionStatus(call);
    }

    @PluginMethod
    public void startTracking(PluginCall call) {
        if (!store.hasSession()) {
            call.reject("Call setAuthSession before startTracking");
            return;
        }
        if (!hasFineLocation()) {
            call.reject("Location permission not granted");
            return;
        }
        store.setTrackingEnabled(true);
        startService();
        JSObject ret = new JSObject();
        ret.put("tracking", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void stopTracking(PluginCall call) {
        store.setTrackingEnabled(false);
        stopService();
        JSObject ret = new JSObject();
        ret.put("tracking", false);
        call.resolve(ret);
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        resolvePermissionStatus(call);
    }

    private void resolvePermissionStatus(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("tracking", store.isTrackingEnabled());
        ret.put("hasSession", store.hasSession());
        ret.put("permission", permissionLabel());
        call.resolve(ret);
    }

    private String permissionLabel() {
        if (!hasFineLocation()) return "denied";
        if (Build.VERSION.SDK_INT >= 29 && !hasBackgroundLocation()) return "whenInUse";
        return "always";
    }

    private boolean hasFineLocation() {
        return ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED
            || ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION)
            == PackageManager.PERMISSION_GRANTED;
    }

    private boolean hasBackgroundLocation() {
        if (Build.VERSION.SDK_INT < 29) return true;
        return ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            == PackageManager.PERMISSION_GRANTED;
    }

    private void startService() {
        Intent intent = new Intent(getContext(), LocationTrackingService.class);
        intent.setAction(LocationTrackingService.ACTION_START);
        ContextCompat.startForegroundService(getContext(), intent);
    }

    private void stopService() {
        Intent intent = new Intent(getContext(), LocationTrackingService.class);
        intent.setAction(LocationTrackingService.ACTION_STOP);
        getContext().startService(intent);
    }
}

File: plugins/background-location/ios/Sources/BackgroundLocationPlugin/BackgroundLocationPlugin.swift
import Foundation
import Capacitor
import CoreLocation

@objc(BackgroundLocationPlugin)
public class BackgroundLocationPlugin: CAPPlugin, CAPBridgedPlugin, CLLocationManagerDelegate {
    public let identifier = "BackgroundLocationPlugin"
    public let jsName = "BackgroundLocation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setAuthSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearAuthSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
    ]

    private let store = SessionStore()
    private lazy var writer = SupabaseLocationWriter(store: store)
    private let manager = CLLocationManager()
    private var pendingPermissionCall: CAPPluginCall?

    public override func load() {
        manager.delegate = self
        manager.pausesLocationUpdatesAutomatically = true
        manager.desiredAccuracy = kCLLocationAccuracyHundredMeters
        manager.distanceFilter = 50
        if store.isTrackingEnabled && store.hasSession {
            beginUpdates()
        }
    }

    @objc func setAuthSession(_ call: CAPPluginCall) {
        guard
            let access = call.getString("accessToken"),
            let userId = call.getString("userId"),
            let url = call.getString("supabaseUrl"),
            let anon = call.getString("supabaseAnonKey")
        else {
            call.reject("accessToken, userId, supabaseUrl, and supabaseAnonKey are required")
            return
        }
        store.saveSession(
            accessToken: access,
            refreshToken: call.getString("refreshToken"),
            userId: userId,
            supabaseUrl: url,
            anonKey: anon
        )
        call.resolve(["stored": true])
    }

    @objc func clearAuthSession(_ call: CAPPluginCall) {
        stopUpdates()
        store.clearSession()
        call.resolve(["cleared": true])
    }

    @objc public override func requestPermissions(_ call: CAPPluginCall) {
        let status = manager.authorizationStatus
        if status == .notDetermined {
            pendingPermissionCall = call
            manager.requestWhenInUseAuthorization()
            return
        }
        if status == .authorizedWhenInUse {
            pendingPermissionCall = call
            manager.requestAlwaysAuthorization()
            return
        }
        call.resolve(statusPayload())
    }

    @objc func startTracking(_ call: CAPPluginCall) {
        guard store.hasSession else {
            call.reject("Call setAuthSession before startTracking")
            return
        }
        let status = manager.authorizationStatus
        guard status == .authorizedAlways || status == .authorizedWhenInUse else {
            call.reject("Location permission not granted")
            return
        }
        store.isTrackingEnabled = true
        beginUpdates()
        call.resolve(["tracking": true])
    }

    @objc func stopTracking(_ call: CAPPluginCall) {
        store.isTrackingEnabled = false
        stopUpdates()
        call.resolve(["tracking": false])
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        call.resolve(statusPayload())
    }

    private func beginUpdates() {
        if manager.authorizationStatus == .authorizedAlways {
            manager.allowsBackgroundLocationUpdates = true
            manager.showsBackgroundLocationIndicator = true
        }
        manager.startUpdatingLocation()
        manager.startMonitoringSignificantLocationChanges()
    }

    private func stopUpdates() {
        manager.stopUpdatingLocation()
        manager.stopMonitoringSignificantLocationChanges()
        manager.allowsBackgroundLocationUpdates = false
    }

    private func statusPayload() -> [String: Any] {
        [
            "tracking": store.isTrackingEnabled,
            "hasSession": store.hasSession,
            "permission": permissionLabel(),
        ]
    }

    private func permissionLabel() -> String {
        switch manager.authorizationStatus {
        case .authorizedAlways: return "always"
        case .authorizedWhenInUse: return "whenInUse"
        case .denied, .restricted: return "denied"
        case .notDetermined: return "prompt"
        @unknown default: return "prompt"
        }
    }

    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        if manager.authorizationStatus == .authorizedWhenInUse, pendingPermissionCall != nil {
            manager.requestAlwaysAuthorization()
            return
        }
        if let call = pendingPermissionCall {
            pendingPermissionCall = nil
            call.resolve(statusPayload())
        }
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard store.isTrackingEnabled, let loc = locations.last else { return }
        guard writer.shouldUpload(lat: loc.coordinate.latitude, lng: loc.coordinate.longitude) else { return }
        writer.upload(lat: loc.coordinate.latitude, lng: loc.coordinate.longitude, accuracyM: loc.horizontalAccuracy)
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        CAPLog.print("BackgroundLocation error: \(error.localizedDescription)")
    }
}
